Sources/MongoSwift/MongoCursor.swift

```diff
@@ -180,59 +205,133 @@ public class MongoCursor<T: Codable>: Sequence, IteratorProtocol {
             }
-            return .success(doc)
         }
+    }
 
-        // We already closed the mongoc cursor, either because we reached the end or encountered an error.
-        guard case let .open(_, connection, client, session) = self.state else {
-            return .failure(ClosedCursorError)
-        }
+    /**
+     * Get the next `T` from the cursor.
+     *
+     * If this cursor is tailable, this method will continue polling until a non-empty batch is returned from the server
+     * or the cursor is closed.
+     *
+     * - Returns:
+     *   An `EventLoopFuture<T?>` evaluating to the next `T` in this cursor, `nil` if the cursor is exhausted,
+     *   or an error if one occurred. If the underlying cursor is tailable, the future will not resolve
+     *   until data is returned (potentially after multiple requests to the server), the cursor is closed, or an error
+     *   occurs.
+     *
+     *   If the future evaluates to an error, it is likely one of the following:
+     *     - `CommandError` if an error occurs while fetching more results from the server.
+     *     - `LogicError` if this function is called after the cursor has died.
+     *     - `LogicError` if this function is called and the session associated with this cursor is inactive.
+     *     - `DecodingError` if an error occurs decoding the server's response.
+     */
+    public func next() -> EventLoopFuture<T?> {
+        return self.client.operationExecutor.execute {
+            try self.lock.withLock {
+                guard self.isAlive else {
+                    throw ClosedCursorError
+                }
+
+                // Keep trying until either the cursor is killed or a notification has been sent by close
+                while self.isAlive && !self.channel.receive() {
+                    if let doc = try self.getNextDocument() {
+                        return doc
+                    }
+                }
 
-        let operation = NextOperation(target: .cursor(self))
-        do {
-            guard let result = try client.executeOperation(operation, using: connection, session: session) else {
                 return nil
             }
-            return .success(result)
-        } catch {
-            return .failure(error)
         }
     }
+
+    /**
+     * Kill this cursor.
+     *
+     * This method MUST be called before this cursor goes out of scope to prevent leaking resources.
+     * This method may be called even if there are unresolved futures created from other `MongoCursor` methods.
+     *
+     * - Returns:
+     *   An `EventLoopFuture` that evaluates when the cursor has completed closing. This future should not fail.
+     */
+    public func kill() -> EventLoopFuture<Void> {
+        return self.client.operationExecutor.execute {
+            self.channel.notify()
+            self.lock.withLock {
+                self.blockingKill()
+            }
+        }
+    }
+}
+
+/// A channel used to notify other threads.
```

Comment:
technically this can only be used to notify one other thread, and only one time, correct? (that seems fine for our purposes, just checking my understanding)

additionally, are we meeting this requirement?
> The NSLock class uses POSIX threads to implement its locking behavior. When sending an unlock message to an NSLock object, you must be sure that message is sent from the same thread that sent the initial lock message. Unlocking a lock from a different thread can result in undefined behavior.
(from [here](https://developer.apple.com/documentation/foundation/nslock))

it seems to me like since `close` can be called from any thread, and also we are running `close`'s body in a thread pool, that we are likely not unlocking from the same thread we called lock in.

Reply:
This is definitely a problem, so opted for a simpler approach using a `NIOAtomic<Bool>` instead of a channel-like thing.